File: v2/lights-out-v2.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <vector>
#include <tuple>
#include <math.h>
#include <queue>
#include <time.h>
#include <algorithm>
#include <mutex>
#include <chrono>
#include <thread>
#include <windows.h>
#include <bitset>

using namespace std;


typedef int MovesTillSolved;
typedef int OptimalMove;
typedef unsigned int State;

typedef tuple<State, OptimalMove> Permutation;
typedef vector<vector<bool>> Matrix;
typedef vector<vector<int>> LookUpMatrix;
typedef vector<State> Rotations;

typedef tuple<State, MovesTillSolved, OptimalMove, State> QueueItem; // state, movesTillSolved, optimalMove, parent
typedef tuple<MovesTillSolved, OptimalMove, State> MemoryItem; // movesTillSolved, optimalMove, parent


// Config settings


const int boardSize = 5;
const bool rotationEnabled = true;
bool isModded = false;

const bool threadingEnabled = false;
const int maxThreads = 5;


//const State stateToSolveFrom2 = 16;
//const State stateToSolveFrom3 = 461;
//const State stateToSolveFrom4 = 25397;
//const State stateToSolveFrom5 = 20051570;


// End Config


// Global Variables
const int totalTiles = boardSize * boardSize;
double maxStates = pow(2, totalTiles);

// Mutex
mutex queueLock;
mutex memoryLock;
mutex outputLock;

State memSize = 0;


unordered_set<State> inQueue;
queue<QueueItem> theQueue; // state, move, optimalMove, parent
unordered_map<State, MemoryItem> memory; // state | move, optimalMove, parent




/* Reverses the binary of a number */
State reverseBinary(State& state) {
	auto bits = std::bitset<totalTiles>(state);

	string debugString = "";

	State reversed = 0;
	int j = 0;
	for (int i = totalTiles - 1; i >= 0; i--) {
		reversed += (bits.test(i)) ? pow(2, j) : 0;

		debugString += (bits.test(i)) ? "1" : "0";
		j++;
	}

	//std::cout << "Original: " << state << " Reversed: " << reversed << endl;

	return reversed;
}


/* Shrinks a 2D Boolean Vector into number */
State shrinkToBinary(Matrix& matrix) {
	State state = 0;

	int index = 0;
	for (vector<bool> row : matrix) {
		for (bool bit : row) {
			state += (bit) ? pow(2, index) : 0;
			index++;
		}
	}

	return state;
}




/* Because reinventing the wheel is stupid, the code below is taken from: https://www.prodevelopertutorial.com/given-an-n-x-n-2d-matrix-rotate-it-by-90-degrees-clockwise-in-c-in-place/ */
State rotateState(Matrix& flipMatrix) {
	size_t size = flipMatrix.size();

	/*Flip Diagonally*/
	for (size_t i = 0; i < size; i++) {
		for (size_t j = i; j < size; j++) {
			swap(flipMatrix[i][j], flipMatrix[j][i]);
		}
	}

	/* flip horizontally  */
	for (size_t i = 0; i < size; i++) {
		reverse(flipMatrix[i].begin(), flipMatrix[i].end());
	}

	return shrinkToBinary(flipMatrix);
}

/* Because reinventing the wheel is stupid, the code below is taken from: https://www.prodevelopertutorial.com/given-an-n-x-n-2d-matrix-rotate-it-by-90-degrees-clockwise-in-c-in-place/ */
void rotateLookup(LookUpMatrix& flipMatrix) {
	size_t size = flipMatrix.size();

	/*Flip Diagonally*/
	for (size_t i = 0; i < size; i++) {
		for (size_t j = i; j < size; j++) {
			swap(flipMatrix[i][j], flipMatrix[j][i]);
		}
	}

	/* flip horizontally  */
	for (size_t i = 0; i < size; i++) {
		reverse(flipMatrix[i].begin(), flipMatrix[i].end());
	}
}


Matrix expandToMatrix(State& state) {
	auto bits = std::bitset<totalTiles>(state);
	Matrix matrix;

	for (int i = 0; i < boardSize; i++) {
		vector<bool> row;

		for (int j = 0; j < boardSize; j++) {
			row.push_back(bits.test(i * boardSize + j));
		};

		matrix.push_back(row);
	}

	return matrix;
}


LookUpMatrix createRotationMatrix(int& numberOfRotations) {
	LookUpMatrix rotationMatrix;

	for (int i = 0; i < boardSize; i++) {
		vector<int> row;

		for (int j = 0; j < boardSize; j++) row.push_back(i * boardSize + j);

		rotationMatrix.push_back(row);
	}


	for (int i = 0; i < numberOfRotations; i++) rotateLookup(rotationMatrix);

	return rotationMatrix;
}



Rotations getRotations(State& state) {
	Rotations rotations;
	rotations.push_back(state);

	Matrix matrix = expandToMatrix(state);
	State rotatedState = rotateState(matrix);

	if (rotatedState == state) return rotations;
	rotations.push_back(rotatedState);

	State reverseState = reverseBinary(state);
	if (reverseState == state) return rotations;

	rotations.push_back(reverseState);
	rotations.push_back(reverseBinary(rotatedState));

	return rotations;
}


State getModdedMask(int& tile) {
	string singleOne = "1";
	string doubleOne = singleOne + singleOne;
	string tripleOne = doubleOne + singleOne;

	int topLeft = 0;
	int topRight = boardSize - 1;
	int bottomLeft = totalTiles - (boardSize);
	int bottomRight = totalTiles - 1;

	string boardSizeMinus2Zeros = std::string(boardSize - 2, '0');
	string boardSizeMinus3Zeros = std::string(boardSize - 3, '0');


	// Handle Corners
	if (tile == topLeft || tile == topRight || tile == bottomLeft || tile == bottomRight) {
		State corner = stoi(doubleOne + boardSizeMinus2Zeros + doubleOne, 0, 2);

		if (tile == topLeft) return (corner << ((boardSize - 2) + ((boardSize - 2) * boardSize)));	// Top left
		if (tile == topRight) return (corner << (totalTiles - (boardSize * 2)));					// Top right
		if (tile == bottomLeft) return (corner << (boardSize - 2));									// Bottom Left
		if (tile == bottomRight) return corner;													// Bottom Right
	}

	State topBottom = stoi(tripleOne + boardSizeMinus3Zeros + tripleOne, 0, 2);
	State leftRight = stoi(doubleOne + boardSizeMinus2Zeros + doubleOne + boardSizeMinus2Zeros + doubleOne, 0, 2);

	// Handle top side clicks
	if (topLeft < tile && topRight > tile) return (topBottom << ((totalTiles - 1) - (tile + boardSize + 1)));

	// Handle left side clicks
	if (tile % boardSize == 0) return (leftRight << ((totalTiles - 1) - (tile + boardSize + 1)));

	// Handle right side clicks
	if (tile % boardSize == boardSize - 1) return (leftRight << ((totalTiles - 1) - (tile + boardSize)));

	// Handle bottom side clicks
	if (bottomLeft < tile && bottomRight > tile) return (topBottom << ((totalTiles - 1) - (tile + 1)));

	// Everything else
	return (stoi(tripleOne + boardSizeMinus3Zeros + tripleOne + boardSizeMinus3Zeros + tripleOne, 0, 2) << ((totalTiles - 1) - (tile + boardSize + 1)));
}


State getNormalMask(int& tile) {
	string singleOne = "1";
	string doubleOne = singleOne + singleOne;
	string tripleOne = doubleOne + singleOne;

	int topLeft = 0;
	int topRight = boardSize - 1;
	int bottomLeft = totalTiles - (boardSize);
	int bottomRight = totalTiles - 1;

	string boardSizeMinus1Zeros = std::string(boardSize - 1, '0');
	string boardSizeMinus2Zeros = std::string(boardSize - 2, '0');


	// Handle Corners
	if (tile == topLeft || tile == topRight || tile == bottomLeft || tile == bottomRight) {
		if (tile == topLeft) return (stoi(doubleOne + boardSizeMinus2Zeros + singleOne, 0, 2) << ((totalTiles - (boardSize + 1))));						// Top left
		if (tile == topRight) return (stoi(doubleOne + boardSizeMinus1Zeros + singleOne, 0, 2) << (totalTiles - (boardSize * 2)));						// Top right
		if (tile == bottomLeft) return stoi(singleOne + boardSizeMinus1Zeros + doubleOne, 0, 2) << (boardSize - 2);										// Bottom Left
		if (tile == bottomRight) return stoi(singleOne + boardSizeMinus2Zeros + doubleOne, 0, 2);														// Bottom Right
	}

	// Handle top side clicks
	if (topLeft < tile && topRight > tile) return (stoi(tripleOne + boardSizeMinus2Zeros + singleOne, 0, 2) << (((boardSize - tile) - 1) + ((boardSize - 2) * boardSize)));

	// Handle left side clicks
	if (tile % boardSize == 0) return (stoi(singleOne + boardSizeMinus1Zeros + doubleOne + boardSizeMinus2Zeros + singleOne, 0, 2) << ((boardSize - 1) + (((boardSize - (tile / boardSize)) - 2) * boardSize)));

	// Handle right side clicks
	if (tile % boardSize == boardSize - 1) return (stoi(singleOne + boardSizeMinus2Zeros + doubleOne + boardSizeMinus1Zeros + singleOne, 0, 2) << (((boardSize - ((tile + 1) / boardSize)) - 1) * boardSize));

	// Handle bottom side clicks
	if (bottomLeft < tile && bottomRight > tile) return (stoi(singleOne + boardSizeMinus2Zeros + tripleOne, 0, 2) << ((totalTiles - 2) - tile));

	// Everything else
	return (stoi(singleOne + boardSizeMinus2Zeros + tripleOne + boardSizeMinus2Zeros + singleOne, 0, 2) << ((totalTiles - 1) - (tile + boardSize)));
}


vector<Permutation> getPermutations(State& state) {
	vector<Permutation> permutations;

	for (int i = 0; i < totalTiles; i++) {
		State mask = (isModded) ? getModdedMask(i) : getNormalMask(i);
		permutations.push_back(make_tuple(state ^ mask, i));
	}

	return permutations;
}


bool checkRotations(State& state) {
	bool rotated = false;
	Rotations rotations = getRotations(state);

	for (State rotation : rotations) { // We can skip the first rotation because it's guarenteed to be the same as the original state

		queueLock.lock();
		bool queueCheck = (inQueue.find(rotation) != inQueue.end());
		queueLock.unlock();

		if (queueCheck) return true;

		memoryLock.lock();
		bool memoryCheck = (memory.find(rotation) != memory.end());
		memoryLock.unlock();

		if (memoryCheck) return true;
	}

	return false;
}


void permutationChecker(Permutation& perm, int& move, State& state) {
	State newState = get<0>(perm);
	OptimalMove optimalMove = get<1>(perm);

	queueLock.lock();
	bool queueFound = inQueue.find(newState) != inQueue.end();
	queueLock.unlock();

	if (queueFound) return;
	if (rotationEnabled) {
		if (checkRotations(newState)) return;
	}

	QueueItem queueItem = make_tuple(newState, move + 1, optimalMove, state);
	queueLock.lock();
	theQueue.emplace(queueItem);
	inQueue.insert(newState);
	queueLock.unlock();

	return;
}




bool keepGoing(double& memoryMax) {

	//queueLock.lock();
	//bool queueEmpty = (theQueue.size() == 0);
	//queueLock.unlock();

	//if (queueEmpty) return false;

	memoryLock.lock();
	bool answer = (memory.size() < memoryMax);
	memoryLock.unlock();

	return answer;
}


void processQueue(double maxMemorySize, int threadID) {
	int processed = 0;

	while (keepGoing(maxMemorySize)) {
		queueLock.lock();

		//cout << theQueue.size() << endl;
		//cout << memory.size() << endl;

		if (theQueue.size() == 0) {
			queueLock.unlock();
			continue;
		}

		QueueItem item = theQueue.front();
		theQueue.pop();

		//inQueue.erase(get<0>(item));
		queueLock.unlock();

		State state = get<0>(item);
		MovesTillSolved move = get<1>(item);
		OptimalMove optimalMove = get<2>(item);
		State parent = get<3>(item);

		memoryLock.lock();
		bool foundItem = memory.find(state) != memory.end();
		memoryLock.unlock();

		if (foundItem) continue;

		MemoryItem memoryItem = make_tuple(move, optimalMove, parent);

		memoryLock.lock();
		memory.insert({ state, memoryItem });
		memoryLock.unlock();

		vector<Permutation> permutations = getPermutations(state);
		for (Permutation perm : permutations) {
			permutationChecker(perm, move, state);
		}

		processed++;
	}

	outputLock.lock();
	std::cout << threadID << " | Thread Finished. Processed: " << processed << "\n";
	outputLock.unlock();
}





int main() {

	//State solveState = setup();
	State startPoint = maxStates - 1;

	string boardSizeString = "Board Size: " + to_string(boardSize);
	boardSizeString += to_string(boardSize);

	string rotationString = " | Rotation: ";
	rotationString += (rotationEnabled) ? "Enabled" : "Disabled";

	string moddedString = " | Modded: ";
	moddedString += (isModded) ? "True" : "False";

	string threadString = "\nThreading: ";
	threadString += (threadingEnabled) ? "Enabled" : "Disabled";

	string maxThreadsString = " | Max Threads: ";
	maxThreadsString += to_string(maxThreads);


	std::cout << "Finding Solution...\nSettings:\n" << boardSizeString << rotationString << moddedString << threadString << maxThreadsString << endl;

	clock_t start = clock();
	theQueue.emplace(make_tuple(startPoint, 0, -1, 0));


	double stoppingPoint = maxStates;

	if (rotationEnabled) {
		if (boardSize == 2) stoppingPoint = 6;
		if (boardSize == 3) stoppingPoint = 140;
		if (boardSize == 4) stoppingPoint = 16456;
		if (boardSize == 5) stoppingPoint = 2099264;
	}

	//cout << maxStates << endl;

	if (threadingEnabled) {
		thread threadList[maxThreads];

		for (int i = 0; i < maxThreads; i++) {
			threadList[i] = thread(processQueue, stoppingPoint, i);
		}

		for (int i = 0; i < maxThreads; i++) {
			threadList[i].join();
		}
	}
	else {
		processQueue(stoppingPoint, 0);
	}


	clock_t end = clock();
	double elapsed = double(end - start);
	double inSeconds = elapsed / 1000;

	std::cout << "Took: " << elapsed << "ms\n";
	std::cout << "In seconds: " << inSeconds << "s\n\n";

	system("pause");

	return 0;
}

File: v2/lights-out-v2.js
"use strict";
import chalk from "chalk";


/*
Brute force 'Solver'
1x1, 2x2, 3x3 and 4x4 are all possible to find a solution for, no matter what the initial state is "quickly".

HOWEVER:
I cannot recommend trying to solve a 5x5 or higher board.
Technically, there are 33.5million possible states for a 5x5 board
Even if you take into account certain states are just rotations of others (which this does), it doesn't help much.
A 5x5 board would still only be brought down to around 2.1million states.

However, one thing that I can say with reasonable confidence, is that because I'm using a breadth first search
when a solution does get found, it will be a solution with the least amount of moves possible.
*/



// Config
const boardSize = 2;
const modded = false;
const rotationEnabled = true;



// Initial setup
const totalTiles = boardSize * boardSize;
const maxMemory = Math.pow(2, Math.pow(boardSize, 2));
const solvedState = maxMemory - 1 >>> 0;

/** @type {number[]} */
const memory = new Array();

const queue = new Map();
// Functions

const nextInQueue = () => { // Use in CPP
	const next = queue.entries().next().value;
	queue.delete(next[0]);

	return next[1];
};


function makeBinaryArray(state) {
	const bits = [];

	let i = state;
	while(i > 0){
		bits.push(i % 2);
		i = Math.floor(i / 2);
	}

	while(bits.length < totalTiles){ // Ensure that it will be the correct length
		bits.push(0);
	}

	// console.log(bits);
	return bits;
}


function reverseBinary(state) {
	const bits = makeBinaryArray(state);
	bits.reverse();

	let decimal = 0;

	// we can use `i` as the power
	for(let i = 0; i < bits.length; i++){
		decimal += bits[i] * Math.pow(2, i);
	}

	return decimal;
}


function shrinkToBinary(matrix) {
	let binaryString = "";

	for(const row of matrix){
		for(const bit of row){
			binaryString += (bit) ? "1" : "0";
		}
	}

	return parseInt(binaryString, 2);
}


function rotate(matrix) {
	const size = matrix.length;
	const layerCount = Math.floor(size / 2);
	for(let layer = 0; layer < layerCount; layer++){
		const first = layer;
		const last = size - first - 1;
		for(let element = first; element < last; element++){
			const offset = element - first;
			const top = matrix[first][element];
			const rightSide = matrix[element][last];
			const bottom = matrix[last][last - offset];
			const leftSide = matrix[last - offset][first];
			matrix[first][element] = leftSide;
			matrix[element][last] = top;
			matrix[last][last - offset] = rightSide;
			matrix[last - offset][first] = bottom;
		}
	}

	return shrinkToBinary(matrix);
}


/** Expand to a 2D matrix */
function expandToMatrix(state) {
	const binary = state.toString(2).padStart(totalTiles, "0").split("");

	const matrix = [];
	for(let i = 0; i < boardSize; i++){
		const row = [];
		for(let j = 0; j < boardSize; j++){
			const index = i * boardSize + j;
			const bit = (binary[index] === "1") ? true : false;
			row.push(bit);
		}

		matrix.push(row);
	}

	return matrix;
}



// /** Produces a list of all possible rotations of a given state */
function getRotations(state) {

	/** @type {number[]} */
	const rotations = [state];


	const rotatedState = rotate(expandToMatrix(state));
	if(rotatedState === state) return rotations; // This shouldn't happen?
	rotations.push(rotatedState);

	rotations.push(reverseBinary(state), reverseBinary(rotatedState));

	return rotations;
}



function getModdedMask(tile) {
	const singleOne = "1";
	const doubleOne = singleOne + singleOne;
	const tripleOne = doubleOne + singleOne;

	const topLeft = 0;
	const topRight = boardSize - 1;
	const bottomLeft = totalTiles - (boardSize);
	const bottomRight = totalTiles - 1;

	const boardSizeMinus2Zeros = "0".repeat(boardSize - 2);
	const boardSizeMinus3Zeros = "0".repeat(boardSize - 3);


	// Handle Corners
	if(tile === topLeft || tile === topRight || tile === bottomLeft || tile === bottomRight){
		const corner = parseInt(doubleOne + boardSizeMinus2Zeros + doubleOne, 2);

		if(tile === topLeft) return corner << (boardSize - 2) + ((boardSize - 2) * boardSize);	// Top left
		if(tile === topRight) return corner << totalTiles - (boardSize * 2);					// Top right
		if(tile === bottomLeft) return corner << boardSize - 2;									// Bottom Left
		if(tile === bottomRight) return corner;													// Bottom Right
	}

	const topBottom = parseInt(tripleOne + boardSizeMinus3Zeros + tripleOne, 2);
	const leftRight = parseInt(doubleOne + boardSizeMinus2Zeros + doubleOne + boardSizeMinus2Zeros + doubleOne, 2);

	// Handle top side clicks
	if(topLeft < tile && topRight > tile) return topBottom << (totalTiles - 1) - (tile + boardSize + 1);

	// Handle left side clicks
	if(tile % boardSize === 0) return leftRight << (totalTiles - 1) - (tile + boardSize + 1);

	// Handle right side clicks
	if(tile % boardSize === boardSize - 1) return leftRight << (totalTiles - 1) - (tile + boardSize);

	// Handle bottom side clicks
	if(bottomLeft < tile && bottomRight > tile) return topBottom << (totalTiles - 1) - (tile + 1);

	// Everything else
	return parseInt(tripleOne + boardSizeMinus3Zeros + tripleOne + boardSizeMinus3Zeros + tripleOne, 2) << (totalTiles - 1) - (tile + boardSize + 1);
}



function getNormalMask(tile) {
	const singleOne = "1";
	const doubleOne = singleOne + singleOne;
	const tripleOne = doubleOne + singleOne;

	const topLeft = 0;
	const topRight = boardSize - 1;
	const bottomLeft = totalTiles - (boardSize);
	const bottomRight = totalTiles - 1;

	const boardSizeMinus1Zeros = "0".repeat(boardSize - 1);
	const boardSizeMinus2Zeros = "0".repeat(boardSize - 2);


	// Handle Corners
	if(tile === topLeft || tile === topRight || tile === bottomLeft || tile === bottomRight){
		if(tile === topLeft) return parseInt(doubleOne + boardSizeMinus2Zeros + singleOne, 2) << totalTiles - (boardSize + 1);							// Top left
		if(tile === topRight) return parseInt(doubleOne + boardSizeMinus1Zeros + singleOne, 2) << totalTiles - (boardSize * 2);	// Top right
		if(tile === bottomLeft) return parseInt(singleOne + boardSizeMinus1Zeros + doubleOne, 2) << boardSize - 2;										// Bottom Left
		if(tile === bottomRight) return parseInt(singleOne + boardSizeMinus2Zeros + doubleOne, 2);														// Bottom Right
	}

	// Handle top side clicks
	if(topLeft < tile && topRight > tile) return parseInt(tripleOne + boardSizeMinus2Zeros + singleOne, 2) << ((boardSize - tile) - 1) + ((boardSize - 2) * boardSize);

	// Handle left side clicks
	if(tile % boardSize === 0) return parseInt(singleOne + boardSizeMinus1Zeros + doubleOne + boardSizeMinus2Zeros + singleOne, 2) << (boardSize - 1) + (((boardSize - (tile / boardSize)) - 2) * boardSize);

	// Handle right side clicks
	if(tile % boardSize === boardSize - 1) return parseInt(singleOne + boardSizeMinus2Zeros + doubleOne + boardSizeMinus1Zeros + singleOne, 2) << ((boardSize - ((tile + 1) / boardSize)) - 1) * boardSize;

	// Handle bottom side clicks
	if(bottomLeft < tile && bottomRight > tile) return parseInt(singleOne + boardSizeMinus2Zeros + tripleOne, 2) << (totalTiles - 2) - tile;

	// Everything else
	return parseInt(singleOne + boardSizeMinus2Zeros + tripleOne + boardSizeMinus2Zeros + singleOne, 2) << (totalTiles - 1) - (tile + boardSize);
}


/** Lists all the possible permutations of a state */
function getPermutations(state) {
	const permutations = [];

	for(let i = 0; i < totalTiles; i++){
		const mask = (modded) ? getModdedMask(i) : getNormalMask(i);
		permutations.push({ perm: state ^ mask, tile: i });
	}

	return permutations;
}

const keepGoing = () => maxMemory > memory.length && queue.size > 0; // Use in CPP
function doQueue(_state = solvedState) {

	/**
	 * Parent +1 = Impossible to permutate to, so we know it's a dead end.
	 * totalTiles + 1 = A tile that obviously cannot be toggled, so we know it's a dead end.
	 * moves: 0 = There are no moves left to make, so we know it's a dead end.
	 *
	 * All 3 of these combined means it'll be easy to indentify later when we solve the puzzle, meaning we know we have reached the end
	**/
	queue.set(solvedState, { state: solvedState, parent: solvedState + 1, movesToSolve: 0, optimalMove: totalTiles + 1 });

	if(rotationEnabled){
		while(queue.size > 0){
			const { state, movesToSolve, parent, optimalMove } = nextInQueue();


			if(memory[state] !== undefined) continue; // This shouldn't happen, but just in case
			memory[state] = { movesToSolve, optimalMove, parent };


			const permutations = getPermutations(state);


			for(const permutation of permutations){
				const { perm, tile } = permutation;
				if(memory[perm] !== undefined || queue.has(perm)) continue;


				const rotations = getRotations(perm);

				let isRotated = false;
				for(let i = 1; i < rotations.length; i++){
					if(memory[rotations[i]] !== undefined || queue.has(rotations[i])){
						isRotated = true;
						break;
					}
				}

				if(isRotated) continue;


				queue.set(perm, { state: perm, movesToSolve: movesToSolve + 1, parent: state, optimalMove: tile });
			}
		}
	} else {
		while(queue.size > 0){
			const { state, movesToSolve, parent, optimalMove } = nextInQueue();


			if(memory[state] !== undefined) continue; // This shouldn't happen, but just in case
			memory[state] = { movesToSolve, optimalMove, parent };


			const permutations = getPermutations(state);


			for(const permutation of permutations){
				const { perm, tile } = permutation;
				if(memory[perm] !== undefined || queue.has(perm)) continue;

				queue.set(perm, { state: perm, movesToSolve: movesToSolve + 1, parent: state, optimalMove: tile });
			}
		}
	}

	console.log("Done!\n\n");
}
console.log("Version 2.0");
console.log("\nFinding solution...");

console.time("Took");
doQueue();
console.timeEnd("Took");

console.log(`\nMemory Size: ${chalk.yellow(memory.filter(i => i !== undefined).length)} | Max Memory: ${chalk.magenta(maxMemory)}`);

const rotationsAreEnabled = (rotationEnabled) ? chalk.green("Enabled") : chalk.red("Disabled");
const isModdedGame = (modded) ? chalk.green("True") : chalk.red("False");

console.log(`Rotation: ${rotationsAreEnabled} | Modded: ${isModdedGame} | Board Size: ${chalk.blue(boardSize)}`);


/*
Version 2.0 Times

3x3 Normal:		4ms (No Rotation)
3x3 Normal:		6ms (With Rotation)

3x3 Modded:		4ms (No Rotation)
3x3 Modded:		6ms (With Rotation)



4x4 Normal:		N/A (Many combinations are impossible to solve)
4x4 Normal: 	N/A (Many combinations are impossible to solve)

4x4 Modded:		595ms (No Rotation)
4x4 Modded:		345ms (With Rotation)



5x5 Normal:		I refuse to test (No Rotation)
5x5 Normal:		7m, 40s, 108ms (With Rotation)

5x5 Modded:		N/A (Many combinations are impossible to solve)
5x5 Modded:		N/A (Many combinations are impossible to solve)

*/